### ISSUE TITLE
Preview build fails consensus checkpoint on Windows

### ISSUE THREAD

**Issue by zathras-crypto** [OmniLayer/omnicore#222]
It's been reported that the latest windows build fails the consensus checkpoint at block 320000:

![consensusfail](https://cloud.githubusercontent.com/assets/5633351/9651144/2404ff9e-5250-11e5-8db2-40a7059a7af6.png)

I've reproduced by testing a ```--startclean``` on mainnet, on Windows, on a build based on our latest (8eb3afa9f12bdebb610b611a89254f639169baaf) and can confirm this does indeed stop parsing with a checkpoint failure at block 320000 when run on mainnet.

```
2015-09-03 05:28:39 VerifyCheckpoint(): consensus hash mismatch - expected 0b510ba8e4ba26e563337053922b31a4f2ec78eac62f73b4b70bad7d026a0b65, received 256f98ab9c32cc643abe94b03b014c984586c535a03ea84311d843966cee0562
2015-09-03 05:28:39 Shutting down due to failed checkpoint for block 320000 (hash 000000000000000015aab005b28a326ade60f07515c33517ea5cb598f28fb7ea)
2015-09-03 05:29:23 Shutdown requested, stop scan at block 320001 of 320004
```

Running the same commit on Linux does not encounter any consensus failures when run with ```--startclean``` on mainnet.

**Comment by zathras-crypto** [comment]
Linux & Windows differ on the math for crowdsale purchase 8fbd96005aba5671daf8288f89df8026a7ce4782a0bb411937537933956b827b.

Linux:
```
parseTransaction(block=315312, 2014-08-12 20:56:54 idx= 1149); txid: 8fbd96005aba5671daf8288f89df8026a7ce4782a0bb411937537933956b827b
	------------------------------
	         version: 0, class B
	            type: 0 (Simple Send)
	        property: 2 (TMSC)
	           value: 30.00000000
update_tally_map(18fMQUB86vgx3WaQnYBdx5D2d5zYN8zKtd, 2=0x2, -3000000000, ttype=0): before=5000000000, after=2000000000
update_tally_map(1CXnRrBm3NezYeAcu555Ubu7swwKsn26dT, 2=0x2, +3000000000, ttype=0): before=0, after=3000000000
INVESTMENT SEND to Crowdsale Issuer: 1CXnRrBm3NezYeAcu555Ubu7swwKsn26dT
logicHelper_CrowdsaleParticipation(): granting via crowdsale to user: 83848836698.67978007 2 (TMSC)
logicHelper_CrowdsaleParticipation(): granting via crowdsale to issuer: 8384883669.86797800 2 (TMSC)
update_tally_map(18fMQUB86vgx3WaQnYBdx5D2d5zYN8zKtd, 2147483696=0x80000030, +8384883669867978007, ttype=0): before=0, after=8384883669867978007
update_tally_map(1CXnRrBm3NezYeAcu555Ubu7swwKsn26dT, 2147483696=0x80000030, +838488366986797800, ttype=0): before=0, after=838488366986797800
eraseMaxedCrowdsale(): ERASING MAXED OUT CROWDSALE from address=1CXnRrBm3NezYeAcu555Ubu7swwKsn26dT, at block 315312 (timestamp: 1407877014), SP: 2147483696 (Test token: 48 : 0x80000030)
eraseMaxedCrowdsale(): 2014-08-12 20:56:54
eraseMaxedCrowdsale(): 1CXnRrBm3NezYeAcu555Ubu7swwKsn26dT : id=2147483696=80000030; prop=2, value= 3133700000000, deadline: 1916-08-16 11:28:32 (1479B99E560)
updateSP(): updated entry for SP 2147483696 successfully
recordTX(8fbd96005aba5671daf8288f89df8026a7ce4782a0bb411937537933956b827b, valid=YES, block= 315312, type= 0, value= 3000000000)
```

Windows:
```
parseTransaction(block=315312, 2014-08-12 20:56:54 idx= 1149); txid: 8fbd96005aba5671daf8288f89df8026a7ce4782a0bb411937537933956b827b
	------------------------------
	         version: 0, class B
	            type: 0 (Simple Send)
	        property: 2 (TMSC)
	           value: 30.00000000
update_tally_map(18fMQUB86vgx3WaQnYBdx5D2d5zYN8zKtd, 2=0x2, -3000000000, ttype=0): before=5000000000, after=2000000000
update_tally_map(1CXnRrBm3NezYeAcu555Ubu7swwKsn26dT, 2=0x2, +3000000000, ttype=0): before=0, after=3000000000
INVESTMENT SEND to Crowdsale Issuer: 1CXnRrBm3NezYeAcu555Ubu7swwKsn26dT
logicHelper_CrowdsaleParticipation(): granting via crowdsale to user: 83848836696.53967730 2 (TMSC)
logicHelper_CrowdsaleParticipation(): granting via crowdsale to issuer: 8384883672.00808077 2 (TMSC)
update_tally_map(18fMQUB86vgx3WaQnYBdx5D2d5zYN8zKtd, 2147483696=0x80000030, +8384883669653967730, ttype=0): before=0, after=8384883669653967730
update_tally_map(1CXnRrBm3NezYeAcu555Ubu7swwKsn26dT, 2147483696=0x80000030, +838488367200808077, ttype=0): before=0, after=838488367200808077
eraseMaxedCrowdsale(): ERASING MAXED OUT CROWDSALE from address=1CXnRrBm3NezYeAcu555Ubu7swwKsn26dT, at block 315312 (timestamp: 1407877014), SP: 2147483696 (Test token: 48 : 0x80000030)
eraseMaxedCrowdsale(): 2014-08-12 20:56:54
eraseMaxedCrowdsale(): 1CXnRrBm3NezYeAcu555Ubu7swwKsn26dT : id=2147483696=80000030; prop=2, value= 3133700000000, deadline: 1916-08-16 11:28:32 (1479B99E560)
updateSP(): updated entry for SP 2147483696 successfully
recordTX(8fbd96005aba5671daf8288f89df8026a7ce4782a0bb411937537933956b827b, valid=YES, block= 315312, type= 0, value= 3000000000)
```

Highlighted:
```diff
-Linux:
-granting via crowdsale to user: 83848836698.67978007 2 (TMSC)
-granting via crowdsale to issuer: 8384883669.86797800 2 (TMSC)
+Windows:
+granting via crowdsale to user: 83848836696.53967730 2 (TMSC)
+granting via crowdsale to issuer: 8384883672.00808077 2 (TMSC)
```

**Comment by dexX7** [comment]
With verbose logging:

- [Ubuntu x64](http://pastebin.com/raw.php?i=0CZcNdNR)
- [Windows x64](http://pastebin.com/raw.php?i=h2CPXhWe)

It starts to differ here:
```diff
 calculateFundraiser: ratio *= satoshi_precision_ = 14420930841687949365000000000
 calculateFundraiser: ratio /= maxCreatable = 156352045478212887156
 calculateFundraiser: issuerTokens_int = issuerTokens_int * precision_ = 1310993712880722669000000000000
 calculateFundraiser: issuerTokens_int *= satoshi_precision_ = 131099371288072266900000000000000000000
-calculateFundraiser: issuerTokens_int /= ratio = 838488366986797800
-calculateFundraiser: createdTokens_int = MAX_INT_8_BYTES - issuerTokens_int = 8384883669867978007
-user:   8384883669867978007
-issuer: 838488366986797800
+calculateFundraiser: issuerTokens_int /= ratio = 838488367200808077
+calculateFundraiser: createdTokens_int = MAX_INT_8_BYTES - issuerTokens_int = 8384883669653967730
+user:   8384883669653967730
+issuer: 838488367200808077
```

**Comment by dexX7** [comment]
I tested different types of the `boost::multiprecision` library, and all of them have issues on Windows:
```c++
BOOST_AUTO_TEST_CASE(big_checked_int128_t)
{
    using boost::multiprecision::checked_int128_t;

    PrintToConsole("checked_int128_t:  max: %s\n",
            boost::lexical_cast<std::string>(std::numeric_limits<checked_int128_t>::max()));

    checked_int128_t num("131099371288072266900000000000000000000");    
    checked_int128_t denom("156352045478212887156");
    checked_int128_t result = num / denom;
    checked_int128_t expected("838488366986797800");

    PrintToConsole("checked_int128_t:  %s / %s = %s (expected: %s)\n",
            boost::lexical_cast<std::string>(num),
            boost::lexical_cast<std::string>(denom),
            boost::lexical_cast<std::string>(result),
            boost::lexical_cast<std::string>(expected));

    BOOST_CHECK_EQUAL(expected, result);
}
```

I tested `int128_t`, `int256_t`, `int1024_t` and `cpp_int`.

All fail on Windows x64:

- https://travis-ci.org/OmniLayer/omnicore/jobs/78651844#L2357-L2357

While it's fine with other OS:

- https://travis-ci.org/OmniLayer/omnicore/jobs/78651843#L1587-L1594

This is critical, because, if simple calculations even within the range of `int128_t` fail, then it may also affect the MetaDEx.

**Comment by zathras-crypto** [comment]
> This is critical, because, if simple calculations even within the range of int128_t fail, then it may also affect the MetaDEx.

Yep, if we can't trust up to ```int128_t``` that's a big problem :(

Can we get away without integer signing?  If so I wonder whether the existing implementations of ```uint160``` and ```uint256``` from ```uint256.h```/```uint256.cpp``` would serve our purposes?

**Comment by dexX7** [comment]
This is an awesome idea!

The following checks all pass:
```c++

BOOST_AUTO_TEST_CASE(uint256_from_uint64_t)
{
    uint64_t number = 103242;

    base_uint<256> a(number);

    BOOST_CHECK_EQUAL(number, a.GetLow64());
}

BOOST_AUTO_TEST_CASE(uint256_add)
{
    uint64_t number_a = 103242;
    uint64_t number_b = 234324;
    uint64_t number_c = number_a + number_b;

    base_uint<256> a(number_a);
    base_uint<256> b(number_b);
    base_uint<256> c = a + b;

    BOOST_CHECK_EQUAL(number_c, c.GetLow64());
}

BOOST_AUTO_TEST_CASE(uint256_sub)
{
    uint64_t number_a = 503242;
    uint64_t number_b = 234324;
    uint64_t number_c = number_a - number_b;

    base_uint<256> a(number_a);
    base_uint<256> b(number_b);
    base_uint<256> c = a - b;

    BOOST_CHECK_EQUAL(number_c, c.GetLow64());
}

BOOST_AUTO_TEST_CASE(uint256_mul)
{
    uint64_t number_a = 503242;
    uint64_t number_b = 234324;
    uint64_t number_c = number_a * number_b;

    base_uint<256> a(number_a);
    base_uint<256> b(number_b);
    base_uint<256> c = a * b;

    BOOST_CHECK_EQUAL(number_c, c.GetLow64());
}

BOOST_AUTO_TEST_CASE(uint256_div)
{
    uint64_t number_a = 9;
    uint64_t number_b = 4;
    uint64_t number_c = number_a / number_b;

    base_uint<256> a(number_a);
    base_uint<256> b(number_b);
    base_uint<256> c = a / b;

    BOOST_CHECK_EQUAL(number_c, c.GetLow64());
}
```

**Comment by zathras-crypto** [comment]
Nice :)  Would you be able to run up a cross-platform test via Travis like you did earlier but instead using uint256?

**Comment by dexX7** [comment]
This is what I'm going to test soon! :)

We might even just kick the boost includes, add a typedef and we're almost good to go.

**Comment by dexX7** [comment]
Alright, as early test I just copied the crowdsale participation into the test:

- [crowdsale_participation_tests.cpp](https://github.com/dexX7/bitcoin/blob/001e0ca62ca65cde5904d0c93340b875068ab810/src/omnicore/test/crowdsale_participation_tests.cpp)


Build ongoing:

- https://travis-ci.org/OmniLayer/omnicore/builds/78739662

The overview will show errors due to OmniJ, which is unrelated. We have to look at the logs and see, whether the Boost tests (via `make check`) pass.

**Comment by dexX7** [comment]
Assuming it does work (and I'm positive), then I'm going to restructure all uses of `cpp_int` and `int128_t`.

It's straight forward for plain `int128_t`, while we have to see what the actual ranges are for `cpp_int`.

It's probably a bit more tricky for the MetaDEx due to the rational number wrapper, but it seems so much easier compared to finding a solution for the general problem here.. (did I say you're the best!? hehe)

**Comment by dexX7** [comment]
YES! The unit tests passed!

**Comment by zathras-crypto** [comment]
> YES! The unit tests passed! 

Awesome, thanks mate that's great!!!

> I'm going to restructure

Thank you bud, really appreciated :)

**Comment by nanomo** [comment]
Issue on 0.0.99-0.10
Windows Build from:
https://github.com/OmniLayer/omnicore/blob/omnicore-0.0.10/doc/metadextestingguide.md

**Comment by dexX7** [comment]
Hey @nanomo, thanks for testing!

The preview hosted on omnichest.info does not include all fixes, which are part of the main branch. At the end of the week we're expecting to publish a release candidate and new binaries.

But just to confirm: do you also see the checkpoint error for block 320000, or another block?

**Comment by nanomo** [comment]
Yes, its in the same block:
015-10-27 09:06:09 parseTransaction(block=319979, 2014-09-10 10:48:53 idx= 492); txid: 1a2aa1abae78cc7102432d2561752e3ddea5bb3c75f28b296209f917a04cdf81
2015-10-27 09:06:09 	------------------------------
2015-10-27 09:06:09 	         version: 0, class B
2015-10-27 09:06:09 	            type: 0 (Simple Send)
2015-10-27 09:06:09 	        property: 28 (SP token: 28)
2015-10-27 09:06:09 	           value: 900.00000000
2015-10-27 09:06:09 update_tally_map(1EHZnScrDaRxnzTMWaBbXmDW4VHafqNnME, 28=0x1C, -90000000000, ttype=0): before=1762181200000000, after=1762091200000000
2015-10-27 09:06:09 update_tally_map(1PtyUTBHEjsSYZsZD7Z7X4DgZ41DhtbUG8, 28=0x1C, +90000000000, ttype=0): before=0, after=90000000000
2015-10-27 09:06:09 recordTX(1a2aa1abae78cc7102432d2561752e3ddea5bb3c75f28b296209f917a04cdf81, valid=YES, block= 319979, type= 0, value= 90000000000)
2015-10-27 09:06:11 VerifyCheckpoint(): consensus hash mismatch - expected 0b510ba8e4ba26e563337053922b31a4f2ec78eac62f73b4b70bad7d026a0b65, received 256f98ab9c32cc643abe94b03b014c984586c535a03ea84311d843966cee0562
_2015-10-27 09:06:11 Shutting down due to failed checkpoint for block 320000 (hash 000000000000000015aab005b28a326ade60f07515c33517ea5cb598f28fb7ea)_
2015-10-27 09:06:11 Shutdown requested, stop scan at block 320001 of 320002
2015-10-27 09:06:11 Exodus balance after initialization: 12935.59467744
2015-10-27 09:06:34 
Omni Core shutdown completed
2015-10-27 09:06:34 Shutdown time: 2015-10-27 09:06:34

**Comment by zathras-crypto** [comment]
Hi guys - I was talking to @CraigSellars on Skype about this very issue yesterday!

> The preview hosted on omnichest.info does not include all fixes

That's correct, but it does include #223 

> Yes, its in the same block # 320000, this its the version Omni Core version v0.10.2.0-b50c70e (64-bit)

The preview version currently on OmniChest is based on d62b0a9 - I just double checked to be sure:

```2015-10-29 01:41:24 Build date: 2015-09-08 02:25:08 +0200, based on commit: d62b0a9```

Could you please update to the preview currently hosted at OmniChest?  You would just need to re-download the file from the OmniChest from the links you noted in the MetaDEx testing guide.

Thanks for testing, please let us know how you get on! :)
Z

**Comment by nanomo** [comment]
from the same place i downloaded the file "omnicore-0.0.9.99-dev-win64-setup" on 9/1/2015 now that file is not available to download, instead i downloaded today 10/29/2015 the file "omnicore-0.0.9.99-dev-win64" and i can confirm the bug is fixed, what a mess of versions :D

and... everithing this work and test to see that i have a balance of 0 OMNI from the bug bounty:
https://github.com/OmniLayer/omnicore/issues/216
LOL

thanks all you guys for the help

**Comment by zathras-crypto** [comment]
Yes, we removed the x64 setup temporarily because the build results were not deterministic and we wanted the build process to generate the same results and be repeatable by anyone.

Great to hear the bug is fixed for you :)  

FYI regarding versions, these are development preview builds of 0.0.10 and so are labelled 0.0.9.99 - they're all previews of the same version (though we can determine the build by the commit ID).  

Sorry to hear you haven't received your bounty yet, I've pinged Patrick and if we haven't heard anything by the next weekly meeting I'll raise it with the team.

Thanks again for testing!
